oxenmq/pubsub.h

```diff
@@ -0,0 +1,172 @@
+ *
+ * FIXME: do we want an unsubscribe, or is expiry / conn management sufficient?
+ *
+ * Type UserData can contain whatever information the user may need at publish time, for example if
+ * the subscription is for logs the subscriber can specify log levels or categories, and the
+ * publisher can choose to send or not based on those.  The UserData type, if provided and non-void,
+ * must be default constructible, must be comparable with ==, and must be movable.
+ */
+template <typename UserData = void>
+class Subscription {
+    static constexpr bool have_user_data = !std::is_void_v<UserData>;
+    using UserData_if_present = std::conditional_t<have_user_data, UserData, detail::no_data_t>;
+    using subdata_t = detail::SubData<UserData>;
+
+    std::unordered_map<ConnectionID, subdata_t> subs;
+    std::shared_mutex _mutex;
+    const std::string description; // description of the sub for logging
+    const std::chrono::milliseconds sub_duration; // extended by re-subscribe
+
+public:
+
+    Subscription() = delete;
+    Subscription(std::string description, std::chrono::milliseconds sub_duration = 30min)
+        : description{std::move(description)}, sub_duration{sub_duration} {}
+
+    // returns true if new sub, false if refresh sub.  throws on error.  `data` will be checked
+    // against the existing data: if there is existing data and it compares `==` to the given value,
+    // false is returned (and the existing data is not replaced).  Otherwise the given data gets
+    // stored for this connection (replacing existing data, if present), and true is returned.
+    bool subscribe(const ConnectionID& conn, UserData_if_present data) {
+        std::unique_lock lock{_mutex};
+        auto expiry = std::chrono::steady_clock::now() + sub_duration;
+        auto [value, added] = subs.emplace(conn, subdata_t{expiry});
+        if (added) {
+            if constexpr (have_user_data)
+                value->second.user_data = std::move(data);
+            return true;
+        }
+
+        value->second.expiry = expiry;
+
+        if constexpr (have_user_data) {
+            // if user_data changed, consider it a new sub rather than refresh, and update
+            // user_data in the mapped value.
+            if (!(value->second.user_data == data)) {
+                value->second.user_data = std::move(data);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // no-user-data version, only available for Subscription<void> (== Subscription without a
+    // UserData type).
+    template <bool enable = !have_user_data, std::enable_if_t<enable, int> = 0>
+    bool subscribe(const ConnectionID& conn) {
+        return subscribe(conn, detail::no_data);
+    }
+
+    // unsubscribe a connection ID.  return the user data, if a sub was present.
+    template <bool enable = have_user_data, std::enable_if_t<enable, int> = 0>
+    std::optional<UserData_if_present> unsubscribe(const ConnectionID& conn) {
```

Comment:
```suggestion
    std::optional<UserData> unsubscribe(const ConnectionID& conn) {
```
because the SFINAE is already ensuring the "if present" part.